### PR TITLE
[Content Encoding] Support serialize for new version of blobProperty

### PR DESCRIPTION
**Pull request by SophieGuo410**
Support serialize for version4 of blobProperty and fix the issue when field is null but after deserialize it's empty.

### CODE REVIEW

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/BlobPropertiesSerDe.java [hunk 1]
```diff
@@ -60,22 +62,22 @@ public static BlobProperties getBlobPropertiesFromStream(DataInputStream stream)
     short accountId = version > VERSION_1 ? stream.readShort() : Account.UNKNOWN_ACCOUNT_ID;
     short containerId = version > VERSION_1 ? stream.readShort() : Container.UNKNOWN_CONTAINER_ID;
     boolean isEncrypted = version > VERSION_2 && stream.readByte() == (byte) 1;
-    String contentEncoding = version > VERSION_3 ? Utils.readIntString(stream) : null;
-    String filename = version > VERSION_3 ? Utils.readIntString(stream) : null;
+    String contentEncoding = version > VERSION_3 ? Utils.readNullableIntString(stream) : null;
+    String filename = version > VERSION_3 ? Utils.readNullableIntString(stream) : null;
```
Comment (jsjtzyy) [inline ambry-messageformat/src/main/java/com/github/ambry/messageformat/BlobPropertiesSerDe.java:66]: minor: looks like we don't have unit test that version = 4 but either contentEncoding or filename is null.

File: ambry-utils/src/main/java/com/github/ambry/utils/Utils.java [hunk 2]
```diff
@@ -149,6 +149,15 @@ public static String readIntString(DataInputStream input) throws IOException {
     return readIntString(input, StandardCharsets.UTF_8);
   }
 
+  public static String readNullableIntString(DataInputStream input) throws IOException {
```
Comment (jsjtzyy) [inline ambry-utils/src/main/java/com/github/ambry/utils/Utils.java:152]: nit: java doc

File: ambry-utils/src/main/java/com/github/ambry/utils/Utils.java [hunk 3]
```diff
@@ -149,6 +149,15 @@ public static String readIntString(DataInputStream input) throws IOException {
...
+    String tmp = readIntString(input, StandardCharsets.UTF_8);
+    if (tmp.length() == 0) {
+      return null;
+    } else {
+      return tmp;
+    }
```
Comment (jsjtzyy) [inline ambry-utils/src/main/java/com/github/ambry/utils/Utils.java:158]: `return tmp.isEmpty() ? null : tmp;`

**Review by jsjtzyy — COMMENTED** [review]
lgtm after minor comments are addressed

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/BlobPropertiesSerDe.java (hunk 1 above)
Comment (SophieGuo410) [reply ambry-messageformat/src/main/java/com/github/ambry/messageformat/BlobPropertiesSerDe.java:66]: I've added an unit test in BlobPropertyTest which set contentEncoding and filename is null.

File: ambry-utils/src/main/java/com/github/ambry/utils/Utils.java (hunk 2 above)
Comment (SophieGuo410) [reply ambry-utils/src/main/java/com/github/ambry/utils/Utils.java:158]: Updated.

File: ambry-utils/src/main/java/com/github/ambry/utils/Utils.java (hunk 3 above)
Comment (SophieGuo410) [reply ambry-utils/src/main/java/com/github/ambry/utils/Utils.java]: Thx, Updated.

File: ambry-messageformat/src/test/java/com/github/ambry/messageformat/BlobPropertiesTest.java [hunk 4]
```diff
@@ -84,10 +84,19 @@ public void basicTest() throws IOException {
     String filenameToExpect = filename;
 
     BlobProperties blobProperties = new BlobProperties(blobSize, serviceId, null, null, false, Utils.Infinite_Time,
+        SystemTime.getInstance().milliseconds(), accountId, containerId, isEncrypted, externalAssetTag, null, null);
+    System.out.println(blobProperties.toString()); // Provide example of BlobProperties.toString()
```
Comment (cgtz) [inline ambry-messageformat/src/test/java/com/github/ambry/messageformat/BlobPropertiesTest.java:88]: is this println intended?

**Review by cgtz — APPROVED** [review]
LGTM

**Review by jsjtzyy — APPROVED** [review]

File: ambry-messageformat/src/test/java/com/github/ambry/messageformat/BlobPropertiesTest.java (hunk 4 above)
Comment (cgtz) [reply ambry-messageformat/src/test/java/com/github/ambry/messageformat/BlobPropertiesTest.java:88]: ah, i see that other tests do it to cover the toString method.

**Pull request merged**